src/library/base/R/zzz.R

```diff
@@ -226,6 +225,8 @@ assign("as.integer", function(x, ...) UseMethod("as.integer"),
        envir = .GenericArgsEnv)
 assign("as.logical", function(x, ...) UseMethod("as.logical"),
        envir = .GenericArgsEnv)
+assign("%*%", function(x, y) UseMethod("%*%"),
+       envir = .GenericArgsEnv)
```

Comment:
IIUC we won't do this,  but rather keep %*% primitive  and use DispatchOrEval() in its C code.
I'm pretty sure the above change would be disruptive,  also a performance killer in some important cases.

Reply:
Thank you @mmaechler for the review!

I reverted this line. Unfortunately, it causes one test to fail. Specifically, the test that checks that none of the symbols in `.ArgsEnv` are generic: https://github.com/r-devel/r-svn/blob/c5f16db96ad76504893a1070d3b46b4093bb49c4/tests/primitives.R#L30-L48

Digging a little, my understanding is that `.GenericArgsEnv` and `.ArgsEnv` are only used for (infrequent) code introspection and in tests, and that the `UseMethod()` call in the `body()` of closures in `.GenericArgsEnv` is not actually evaluated during normal code execution.

Note that the other primitive generics that dispatch to Ops (and similarly are performance sensitive) are also assigned to `.GenericArgsEnv` in the same manner (as a closure with a `UseMethod()` body: 
https://github.com/r-devel/r-svn/blob/e22e2a7c353d27134fcfcbbaee9c9c64952f772e/src/library/base/R/zzz.R#L192-L193

Any suggestion on how to proceed? (e.g., special case an exception for `%*%` in the linked test, or bind a dummy closure to `%*%` in `.GenericArgsEnv`).

Reply:
Thank you for the review @DavisVaughan! 

Responses in line:

> I think that when we submit this to base R, it should be in the form of 4 small PRs that will be easier for them to digest:
> 
>   * Adding S3 generic support to `%*%` (S7 agnostic)
> 
>   * Making `@` support `DispatchOrEval()` as it is documented (S7 agnostic)
> 
>   * Making `inherits()` aware of S7 objects
> 
>   * Making `Ops` aware of S7 objects
> 
> 
> Submitting all 4 at once in the same patch would probably slow things down a lot.

Good idea! I will submit this as 4 separate patches.


> I would also probably try to avoid changing existing code as much as possible. Like the changes to `do_AT()` could probably be simplified greatly to just focus on `DispatchOrEval()`. Ignoring everything else for now seems like the best way to actually get the changes into R.

I've updated `do_AT()` to minimize the size of the diff by collecting all the new behavior into an `if` branch that returns early from the function, and leaving as much as possible of the existing code as-is. I've also removed the default fallback of `getAttrib()` in `@` and `setAttrib()` in `@<-`, leaving an error message instead. 
 
> Some other things I noticed that should be updated:
> 
>   * [x]  Move to S7 not R7

Completed via https://github.com/RConsortium/OOP-WG/pull/276


>   * [x]  Update styling to match base R (match `if(` spacing, match 8 space tabs). I think we might need to get your RStudio IDE (or whatever IDE you use) synced with what base R uses smile , I have a comment below about the specific settings i used that worked for me.

Regarding indentation whitespace, I've (manually) applied the following rule to the patch:
- each additional indentation level is 4 spaces
- strings of 8 leading spaces are replaced by a tab character.

Regarding `if (` vs `if(`, I'm not sure there is much of any convention. 
```bash
$ grep -roh 'if (' src/main | wc -l
9723
$ grep -roh 'if(' src/main | wc -l
4009
```
The current state of the code reflects my best effort to adapt to the local style, but even that is difficult to reliably infer what is preferred.

For posterity, this was also discussed in:
- https://github.com/r-devel/r-svn/pull/100#discussion_r1103884478
- https://github.com/r-devel/r-svn/pull/100#discussion_r1103881078

If there is still appetite for discussions about formatting, I would direct efforts towards
https://github.com/r-devel/r-svn/pull/101


>   * [x]  Update `@<-` docs to not be so S4 specific. Probably need to update `@` docs to mention that it works on non-S4 objects.

Done.

>   * [x]  Update `InternalGenerics` docs to know about `@` being an internal generic https://github.com/wch/r-source/blob/37e2fd10b9ec771040e755691e4d10f7068c3431/src/library/base/man/InternalMethods.Rd#L56

Done.
 
>   * [x]  Update docs of `%*%` to mention that it now supports S3 methods (The `Details` section says "This operator is S3 generic but not S3 generic...")

Done.
 
>   * [ ]  Look into updating `getGroupMembers()` to be aware of `%*%`? That might not be possibly since it isn't really a true "group op"?

For posterity, this was discussed in the most recent OOP working group meeting. Three alternatives discussed were:
 - Defining a new `Matrix` group, (maybe with `t()` `crossprod()` and `tcrossprod()`).
 - Adding `%*%` to the `Arith` group.
 - Doing nothing.

We decided that the group definitions in the methods package will not be updated as part of this patch.